cpp/include/cudf/lists/detail/gather.cuh

```diff
@@ -0,0 +1,219 @@
+ * @param source_column View into the column to gather from
+ * @param gather_map Iterator access to the gather map for `source_column`
+ * map
+ * @param gather_map_size Size of the gather map.
+ * @param nullify_out_of_bounds Nullify values in `gather_map` that are out of bounds
+ * @param stream CUDA stream on which to execute kernels
+ * @param mr Memory resource to use for all allocations
+ *
+ * @returns std::pair containing two columns of data.  the compacted offsets
+ *          for the child column and the base (original) offsets from the source
+ *          column that matches each of those new offsets.
+ *
+ */
+template <bool NullifyOutOfBounds, typename MapItType>
+std::pair<std::unique_ptr<column>, std::unique_ptr<column>> make_gather_offsets(
+  cudf::lists_column_view const& source_column,
+  MapItType gather_map,
+  size_type gather_map_size,
+  cudaStream_t stream                 = 0,
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource())
+{
+  // size of the gather map is the # of output rows
+  size_type output_count = gather_map_size;
+  size_type offset_count = output_count + 1;
+
+  // offsets of the source column
+  size_type const* src_offsets{source_column.offsets().data<size_type>()};
+  int src_size = source_column.size();
+
+  // outgoing offsets.  these will persist as output from the entire gather operation
+  auto dst_offsets_c = cudf::make_fixed_width_column(
+    data_type{INT32}, offset_count, mask_state::UNALLOCATED, stream, mr);
+  mutable_column_view dst_offsets_v = dst_offsets_c->mutable_view();
+  size_type* dst_offsets            = dst_offsets_v.data<size_type>();
+
+  // generate the compacted outgoing offsets.
+  auto count_iter = thrust::make_counting_iterator<size_type>(0);
+  thrust::plus<size_type> sum;
+  thrust::transform_exclusive_scan(
+    rmm::exec_policy(stream)->on(stream),
+    count_iter,
+    count_iter + offset_count,
+    dst_offsets,
+    [gather_map, output_count, src_offsets, src_size] __device__(size_type index) -> size_type {
+      // last offset index is always the previous offset_index + 1, since each entry in the gather
+      // map represents a virtual pair of offsets
+      size_type offset_index = index < output_count ? gather_map[index] : gather_map[index - 1] + 1;
+
+      // if this is an invalid index, this will be a NULL list
+      if (NullifyOutOfBounds && ((offset_index < 0) || (offset_index >= src_size))) { return 0; }
+
+      // the length of this list
+      return src_offsets[offset_index + 1] - src_offsets[offset_index];
+    },
+    0,
+    sum);
+
+  // for each span of offsets (each output row) we need to know the original offset value to build
+  // the gather map for the next level.  this data is temporary and will only persist until the
+  // next level of recursion is done using it.  This way, even if we have a hierarchy 1000 levels
+  // deep, we will only ever have at most 2 of these extra columns in memory.
```

Comment:
Is there a way to reuse the memory instead of allocate/free at each level? I suppose nested lists could have different sized offsets columns so there may be no easy way to find a common reusable size.

Reply:
Right.  One potential optimization is to only free/reallocate if the new buffer needs to be larger.  I'll poke at this a bit.

Reply:
Ok, added this optimization.  In practice it doesn't hit very often as it requires that the # of rows at each successive level of nesting is going down, but it can happen with large gather maps and small nested columns.

I also cleaned up and centralized a bunch of stuff to go along with this.  The cudaMemcpySync() calls now happen in make_gather_data() directly instead of having to be computed by the caller.  And make_gather_data() returns the entire gather_data struct rather than returning the pieces.   Makes the higher level functions cleaner.